File: src/main/java/org/pragmatica/example/ticketing/shared/Money.java
package org.pragmatica.example.ticketing.shared;

import org.pragmatica.lang.Cause;
import org.pragmatica.lang.Result;
import org.pragmatica.lang.Verify;
import org.pragmatica.lang.parse.Number;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Locale;


/// Non-negative monetary amount in minor units. A record's canonical constructor is unavoidably
/// public and cannot return a `Result`, so the non-negativity invariant is held at every
/// construction boundary instead: `money(...)` validates, `plus`/`sum` add non-negatives, and
/// `scaledByPercent` takes a positive `Percent`. No reachable path constructs a negative `Money`.
public record Money(long amountMinor, Currency currency) {
    public enum Currency {
        USD,
        EUR,
        GBP;
        public static Result<Currency> currency(String raw) {
            return Result.lift(Error.unknownCurrency(raw),
                               () -> Currency.valueOf(raw.trim().toUpperCase(Locale.ROOT)));
        }
    }

    public sealed interface Error extends Cause {
        record MalformedAmount(String raw) implements Error {
            @Override
            public String message() {
                return "Amount is malformed: " + raw;
            }
        }

        record NegativeAmount(long amountMinor) implements Error {
            @Override
            public String message() {
                return "Amount must not be negative: " + amountMinor;
            }
        }

        record UnknownCurrency(String raw) implements Error {
            @Override
            public String message() {
                return "Unknown currency: " + raw;
            }
        }

        record CurrencyMismatch(Currency left, Currency right) implements Error {
            @Override
            public String message() {
                return "Currency mismatch: " + left + " vs " + right;
            }
        }

        static Error malformedAmount(String raw) {
            return new MalformedAmount(raw);
        }

        static Error negativeAmount(long amountMinor) {
            return new NegativeAmount(amountMinor);
        }

        static Error unknownCurrency(String raw) {
            return new UnknownCurrency(raw);
        }

        static Error currencyMismatch(Currency left, Currency right) {
            return new CurrencyMismatch(left, right);
        }
    }

    public static Result<Money> money(String amount, String currency) {
        return Result.all(minorUnits(amount),
                          Currency.currency(currency))
                     .map(Money::new);
    }

    public Result<Money> plus(Money other) {
        return Verify.ensure(other.currency(),
                             Verify.Is::equalTo,
                             currency,
                             mismatch(other))
                     .map(_ -> sum(other));
    }

    // A positive `Percent` applied to a non-negative amount stays non-negative, so the result needs
    // no further check. This is the only way to scale a price: a raw long can no longer mint a
    // negative `Money`.
    public Money scaledByPercent(Percent percent) {
        return new Money(scaledMinor(percent.value()), currency);
    }

    public String render() {
        return BigDecimal.valueOf(amountMinor)
                         .movePointLeft(2)
                         .toPlainString() + " " + currency;
    }

    private static Result<Long> minorUnits(String amount) {
        return Number.parseBigDecimal(amount)
                     .mapError(_ -> Error.malformedAmount(amount))
                     .flatMap(Money::toMinorUnits)
                     .flatMap(Money::rejectNegative);
    }

    private static Result<Long> toMinorUnits(BigDecimal amount) {
        return Result.lift(Error.malformedAmount(amount.toPlainString()),
                           () -> amount.movePointRight(2)
                                       .longValueExact());
    }

    private static Result<Long> rejectNegative(long minor) {
        return Verify.ensure(minor, Verify.Is::nonNegative, Error.negativeAmount(minor));
    }

    private Cause mismatch(Money other) {
        return Error.currencyMismatch(currency, other.currency());
    }

    private Money sum(Money other) {
        return new Money(amountMinor + other.amountMinor(), currency);
    }

    private long scaledMinor(long percent) {
        return BigDecimal.valueOf(amountMinor)
                         .multiply(BigDecimal.valueOf(percent))
                         .divide(BigDecimal.valueOf(100),
                                 0,
                                 RoundingMode.HALF_UP)
                         .longValue();
    }
}

File: fsharp/src/Ticketing/Shared/Money.fs
namespace Ticketing.Shared

open System
open System.Globalization
open Pragmatica

[<RequireQualifiedAccess>]
type Currency =
    | USD
    | EUR
    | GBP

[<RequireQualifiedAccess>]
type MoneyError =
    | MalformedAmount of raw: string
    | NegativeAmount of amountMinor: int64
    | UnknownCurrency of raw: string
    | CurrencyMismatch of left: Currency * right: Currency

    interface Cause with
        member this.Message =
            match this with
            | MoneyError.MalformedAmount raw -> $"Amount is malformed: {raw}"
            | MoneyError.NegativeAmount amountMinor -> $"Amount must not be negative: {amountMinor}"
            | MoneyError.UnknownCurrency raw -> $"Unknown currency: {raw}"
            | MoneyError.CurrencyMismatch(left, right) -> $"Currency mismatch: {left} vs {right}"

[<RequireQualifiedAccess>]
module Currency =
    let parse (raw: string) : Result<Currency, Cause> =
        match raw.Trim().ToUpperInvariant() with
        | "USD" -> Ok Currency.USD
        | "EUR" -> Ok Currency.EUR
        | "GBP" -> Ok Currency.GBP
        | _ -> Error(MoneyError.UnknownCurrency raw :> Cause)

    let name (currency: Currency) = string currency

/// Non-negative monetary amount in minor units. The Java record's canonical constructor is
/// unavoidably public, so the invariant is held there by auditing every construction boundary; in
/// F# the constructor itself is private — no reachable path constructs a negative `Money`, and the
/// only ways in (`parse`, `fromMinor`) return a `Result`.
type Money =
    private
        { AmountMinor: int64
          Currency: Currency }

[<RequireQualifiedAccess>]
module Money =
    let amountMinor (money: Money) = money.AmountMinor

    let currency (money: Money) = money.Currency

    let fromMinor (amountMinor: int64) (currency: Currency) : Result<Money, Cause> =
        amountMinor
        |> Verify.ensure Verify.Is.nonNegative (MoneyError.NegativeAmount amountMinor)
        |> Result.map (fun minor -> { AmountMinor = minor; Currency = currency })

    let private minorUnits (amount: string) : Result<int64, Cause> =
        match Decimal.TryParse(amount, NumberStyles.Number, CultureInfo.InvariantCulture) with
        | false, _ -> Error(MoneyError.MalformedAmount amount :> Cause)
        | true, value ->
            let minor = value * 100m

            // Java's `movePointRight(2).longValueExact()`: sub-cent precision or an overflowing
            // amount is malformed, not silently rounded.
            if Decimal.Truncate minor <> minor
               || minor > decimal Int64.MaxValue
               || minor < decimal Int64.MinValue then
                Error(MoneyError.MalformedAmount amount :> Cause)
            elif minor < 0m then
                Error(MoneyError.NegativeAmount(int64 minor) :> Cause)
            else
                Ok(int64 minor)

    let parse (amount: string) (currency: string) : Result<Money, Cause> =
        result {
            let! minor = minorUnits amount
            and! parsedCurrency = Currency.parse currency
            return { AmountMinor = minor; Currency = parsedCurrency }
        }

    let plus (other: Money) (money: Money) : Result<Money, Cause> =
        if money.Currency = other.Currency then
            Ok { money with AmountMinor = money.AmountMinor + other.AmountMinor }
        else
            Error(MoneyError.CurrencyMismatch(money.Currency, other.Currency) :> Cause)

    /// A positive `Percent` applied to a non-negative amount stays non-negative, so the result needs
    /// no further check. This is the only way to scale a price: a raw int64 can no longer mint a
    /// negative `Money`.
    let scaledByPercent (percent: Percent) (money: Money) : Money =
        let scaled =
            decimal money.AmountMinor * decimal (Percent.value percent) / 100m

        { money with AmountMinor = int64 (Math.Round(scaled, 0, MidpointRounding.AwayFromZero)) }

    let render (money: Money) =
        let amount = (decimal money.AmountMinor / 100m).ToString("0.00", CultureInfo.InvariantCulture)
        $"{amount} {Currency.name money.Currency}"
